language/vm/src/compatibility.rs

```diff
@@ -101,19 +98,32 @@ impl Compatibility {
         // we can remove/change a friend function if the function is not used by any module in the
         // friend list. But for simplicity, we decided to go to the more restrictive form now and
         // we may revisit this in the future.
-        for (func_sig, old_vis) in &old_module.exposed_functions {
-            let new_vis_opt = new_module.exposed_functions.get(func_sig);
-            let is_compatible = match (old_vis, new_vis_opt) {
-                (Visibility::Public, Some(Visibility::Public)) => true,
+        for (name, old_func) in &old_module.exposed_functions {
+            let new_func = match new_module.exposed_functions.get(name) {
+                Some(new_func) => new_func,
+                None => {
+                    struct_and_function_linking = false;
+                    continue;
+                }
+            };
+            let is_vis_compatible = match (old_func.visibility, new_func.visibility) {
+                (Visibility::Public, Visibility::Public) => true,
                 (Visibility::Public, _) => false,
-                (Visibility::Script, Some(Visibility::Script)) => true,
+                (Visibility::Script, Visibility::Script) => true,
                 (Visibility::Script, _) => false,
-                (Visibility::Friend, Some(Visibility::Public)) => true,
-                (Visibility::Friend, Some(Visibility::Friend)) => true,
+                (Visibility::Friend, Visibility::Public)
+                | (Visibility::Friend, Visibility::Friend) => true,
                 (Visibility::Friend, _) => false,
                 (Visibility::Private, _) => unreachable!("A private function can never be exposed"),
             };
-            if !is_compatible {
+            if !is_vis_compatible
+                || old_func.parameters != new_func.parameters
+                || old_func.return_ != new_func.parameters
```

Comment:
```suggestion
                || old_func.return_ != new_func.return_
```

Reply:
In fact, I am surprised that only two tests failed... This is a coincidence that in most `republish_*.mvir` tests, functions have no params and no returns.